### ISSUE TITLE
Phi3Vision performs well on training/eval dataset but not well in generation/inference post training

### ISSUE THREAD

**Issue by hm-ca** [microsoft/Phi-3CookBook#140]
Hi Team,

I trained Phi3-Vision on my short video captioning dataset. 
The training went very well (after multiple attempts), and the model reached a good performance point (i.e.: low train and validation loss as well as a high Bert similarity score on validation dataset comparing the generated caption to ground truth captions) (plz see screenshot below)

However post training, when using the trained model checkpoints to generate captions for videos, I'm getting captions that are mostly incorrect even for those samples from the training and validation datasets that scored very high during training eval!

For inference, I ensured I use the same data prep techniques I used for training (minus teacher forcing) and I use the generate function with  greedy decoding, nucleus sampling, etc... but don't seem to get consistent good results (~15% match compared to ~82%during training)

What am I missing? any help is really appreciated!

<img width="1115" alt="Screenshot 2024-08-12 at 1 02 29 PM" src="https://github.com/user-attachments/assets/e3093124-b097-49a3-98ff-8d47ffc17e34">

**Comment by leestott** [comment]
Hi @hm-ca 

The discrepancy between training and inference performance can be frustrating. Here are a few potential areas to investigate:

Overfitting: Despite good training and validation metrics, the model might be overfitting to the training data. Try using techniques like dropout, data augmentation, or regularization to mitigate this.

Inference Settings: Ensure that the inference settings (e.g., decoding strategies) are optimal. Sometimes, tweaking parameters like temperature, top-k, or top-p sampling can improve results.

Data Distribution: Check if the distribution of your training data matches the inference data. If there’s a significant difference, the model might struggle to generalize.

Post-Processing: Implement post-processing steps to refine the generated captions. This can include filtering out nonsensical outputs or using additional models to enhance quality.

**Comment by hm-ca** [comment]
Thanks @leestott for the suggestions.

On Overfitting, I actually had to implement many regularization techniques such as dropout, data augmentation, etc... to get to this level of performance shown in the original message.

Where I'm struggling is on inference post training where the model is generating very coherent captions for the videos however they are not correct 80% of the times (hallucination?). I tried different decoding techniques like beam search, nucleus sampling, other samplings, etc... but not seeing much improved inferencing performance.
Btw, given my video captioning task and long sequences, beam search is consuming a ton of memory even with 1  x H100-80GB, I can barely get it to run with num_beams=3

I'll look closely into data distribution, but the thing is the model is performing extremely well during the validation step of training (different dataset split from training), it's just not performing equally even for the **same checkpoint and the same data** post training, which is concerning.

Been trying to mimic the same decoding techniques the HF Trainer uses during the training eval stage (mostly greedy decoding) including data collation, however so far that approach has yielded no material inference improvements either :|

Appreciate all input/pointers...

Best!
